arangod/Aql/QuerySnippet.cpp

```diff
@@ -275,9 +278,30 @@ void QuerySnippet::serializeIntoBuilder(
         auto permuter = localExpansions.find(current);
         if (permuter != localExpansions.end()) {
           auto collectionAccessingNode = dynamic_cast<CollectionAccessingNode*>(clone);
-          TRI_ASSERT(collectionAccessingNode != nullptr);
-          // Get the `i` th shard
-          collectionAccessingNode->setUsedShard(*std::next(permuter->second.begin(), i));
+          if (collectionAccessingNode != nullptr) {
+            // we guarantee that we only have one collection here
+            TRI_ASSERT(permuter->second.size() == 1);
+            std::string const& cName = collectionAccessingNode->collection()->name();
+            // Get the `i` th shard
+            // TODO optimize cName
+            collectionAccessingNode->setUsedShard(*std::next(permuter->second.at(cName).begin(), i));
+          } else {
+            // GraphNodes are not CollectionAccessingNodes! (they handle multiple collections)
+            auto graphNode = dynamic_cast<GraphNode*>(clone);
+            if (graphNode != nullptr) {
+              if (graphNode->isDisjoint()) {
+                TRI_ASSERT(graphNode != nullptr);
+                TRI_ASSERT(graphNode->isDisjoint());
+                // we've found a disjoint smart graph node, now add the `i` th shard for used collections
+                for (auto const& myExp : permuter->second) {
+                  std::string const& cName = myExp.first;
+                  graphNode->addCollectionToShard(cName, *std::next(myExp.second.begin(), i));
+                }
+              }
+            } else {
+              TRI_ASSERT(false);
```

Comment:
For production please throw an error here after the assert.
It should never happen, but if it happens for some reason there is a chance we end up in UB. later.
Better avoid it.